### DISCUSSION TITLE
Packing ngsolve with PyInstaller

### DISCUSSION THREAD

**Discussion by ch1lln** [pyinstaller/pyinstaller#7722] [Help] [answered]
Dear community,

I am trying to package ngsolve (https://pypi.org/project/ngsolve/, https://ngsolve.org/) with the newest version of  PyInstaller on Python 3.8.
NGSolve is distributed only via PIP not conda.

So if you create a test script with just `import ngsolve` then it will already crash when executing.

>  ngsolve.ngslib not found in PYZ
> Traceback (most recent call last):
>   File "test.py", line 5, in <module>
>   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
>   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
>   File "ngsolve\__init__.py", line 36, in <module>
>     from .ngslib import __version__, ngstd, bla, la, fem, comp, solve
>   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 657, in _load_unlocked
>   File "<frozen importlib._bootstrap>", line 556, in module_from_spec
>   File "<frozen importlib._bootstrap_external>", line 1166, in create_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
> ImportError: DLL load failed while importing ngslib: A dynamic link library (DLL) initialization routine failed

Clearly the error lies in the submodule ngslib which cannot be loaded for some reason.

The wheel of ngsolve provides a ngslib.pyd & ngslib.lib. ngslib.pyd refers to other .DLLs provided by a dependency package of ngsolve which is netgen-mesher.

I've tried: 
- --collect-all for netgen & ngsolve
- --hidden-import ngsolve.ngslib
- custom hook for all submodules of ngslib (hiddenimports = ['ngsolve.ngslib','ngsolve.ngslib.ngstd', 'ngsolve.ngslib.bla', 'ngsolve.ngslib.la', 'ngsolve.ngslib.fem', 'ngsolve.ngslib.comp', 'ngsolve.ngslib.solve'])
- copying all related DLLs and pyds manually into the dist folder

The error never changes and I am out of ideas how to troubleshoot this error further.
I'd be very grateful for any experience with this package or some hints into the right direction

**Comment by rokm** [comment, marked as answer]
>  NGSolve is distributed only via PIP not conda.

Since you explicitly mention this, is your python a python.org python or anaconda python?

> copying all related DLLs and pyds manually into the dist folder

What DLLs did you copy, exactly? 

`ngsolve` seems to depend on `mkl`, `intel-openmp` and `tbb`, all of which seem to install their DLLs in `<venv>\Library\bin`. MKL DLLs perform delayed/dynamic loading, so copying just `mkl_rt.2.dll` and `mkl_intel_thread.2.dll` is not enough. If you have no experience with MKL and don't know which DLLs exactly you need (which I suspect might also on the CPU where the program is running...), it is probably safest to copy them all...

EDIT: and you need to copy them in the top-level application directory (without `Library\bin` subdir). The latest release of `ngsolve` still tries to find `mkl_rt.2.dll` in directory that is outside of the frozen application bundle (see [here](https://github.com/NGSolve/ngsolve/blob/33bcde0ca62a460694cb2927a3a8a78b3dfeb521/python/__init__.py#L31-L34); they changed that in https://github.com/NGSolve/ngsolve/commit/dc76db7a09e7d6d4490db93e40f52c7ded75ff12, although I haven't checked if for better or worse...) - but our `ctypes` run-time override performs fall back to top-level application directory, and loads `mkl_rt.2.dll` from there.

**Reply by ch1lln** [reply]
Awesome thank you for the very detailed answer.

> Since you explicitly mention this, is your python a python.org python or anaconda python?

I am using a conda-environment (miniforge) but not anaconda.

> What DLLs did you copy, exactly?
I copied DLLs from netgen and ngsolve because I erroneously thought that was the problem.

But as you correctly said the problem are the MKL DLLs. I copied all mkl DLLs from Library\bin into the top-level app directory and it works now.

I wrote this quick hook (hook-ngsolve.py) to just get all of them automatically:


```
from pathlib import Path

from PyInstaller import compat
from PyInstaller import log as logging

logger = logging.getLogger(__name__)
libdir = Path(compat.base_prefix) / "Library/bin"
mkllib = list(libdir.glob("mkl_*.dll"))
lib_mp5md = libdir / "libiomp5md.dll"
if lib_mp5md.exists():
   mkllib.append(lib_mp5md)
   
if len(mkllib) > 0:
   logger.info(f"Grabbing all {len(mkllib)} MKL DLLs")
   binaries = [(str(file.resolve()), ".") for file in mkllib]

```
